File: benchmarks/complex/pmontecarlo.cpp
#include <cstdio>
#include <thread>
#include <atomic>
static long sample(long base) {
    long seed = base * 2654435761L + 1, count = 0;
    for (long i = 0; i < 25000000L; i++) {
        seed = (seed * 1103515245L + 12345) % 2147483648L;
        long x = seed % 32768;
        seed = (seed * 1103515245L + 12345) % 2147483648L;
        long y = seed % 32768;
        if (x * x + y * y <= 1073676289L) count++;
    }
    return count;
}
int main() {
    std::atomic<long> hits{0};
    std::thread ts[4];
    for (long b = 0; b < 4; b++) ts[b] = std::thread([&hits, b]{ hits.fetch_add(sample(b)); });
    for (auto& t : ts) t.join();
    printf("%ld\n", hits.load());
}

File: benchmarks/complex/pmontecarlo.rs
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;
use std::thread;
fn sample(base: i64) -> i64 {
    let mut seed = base * 2654435761 + 1;
    let mut count = 0i64;
    for _ in 0..25000000i64 {
        seed = (seed * 1103515245 + 12345) % 2147483648;
        let x = seed % 32768;
        seed = (seed * 1103515245 + 12345) % 2147483648;
        let y = seed % 32768;
        if x * x + y * y <= 1073676289 { count += 1; }
    }
    count
}
fn main() {
    let hits = Arc::new(AtomicI64::new(0));
    let mut hs = vec![];
    for b in 0..4i64 {
        let h = hits.clone();
        hs.push(thread::spawn(move || { h.fetch_add(sample(b), Ordering::Relaxed); }));
    }
    for t in hs { t.join().unwrap(); }
    println!("{}", hits.load(Ordering::Relaxed));
}
